Mathlib/Topology/Compactification/OnePointEquiv.lean

```diff
@@ -0,0 +1,166 @@
+  ite (r ≠ 0) (a / r) ∞
+
+/-- Uncurried version of `divOnePoint`, with nonzeroness assumption. -/
+noncomputable def divOnePoint' {K : Type} [DivisionRing K]
+  (u : {v : Fin 2 → K // v ≠ 0}) : OnePoint K :=
+  if u.1 1 ≠ 0 then some (u.1 0 / u.1 1) else ∞
+
+/-- Notation `÷` showing that `div_onePoint` is distinct from
+the ordinary division `/` of a `DivisionRing`. -/
+infix:50 " ÷ " => divOnePoint
+
+/-- `div_onePoint` can be lifted to the projective line (see `divSlope`.) -/
+lemma divOnePoint_lifts {K : Type} [Field K] (a b : {v : Fin 2 → K // v ≠ 0})
+    (h : ∃ c : Kˣ, (fun m : Kˣ ↦ m • b.1) c = a.1) :
+    (fun u ↦ u.1 0 ÷ u.1 1) a = (fun u ↦ u.1 0 ÷ u.1 1) b := by
+  obtain ⟨c,hc⟩ := h
+  simp_all only
+  rw [← hc]; unfold divOnePoint; simp only [ne_eq, Fin.isValue, Pi.smul_apply, ite_not]
+  split_ifs with hbc hb hb
+  · rfl
+  · simp_all only [ne_eq, OnePoint.infty_ne_coe]
+    apply hb;exact (Units.mul_right_eq_zero c).mp hbc
+  · rw [hb] at hbc;simp at hbc
+  · apply congrArg some
+    field_simp
+    show c * b.1 0 * b.1 1 = b.1 0 * (c * b.1 1)
+    ring
+
+/-- Equivalence between the projective line and the one-point extension. -/
+noncomputable def divSlope {K : Type} [Field K] (p : ℙ K (Fin 2 → K)) : OnePoint K :=
+  Quotient.lift (fun u => divOnePoint (u.1 0) (u.1 1)) divOnePoint_lifts p
+
+/-! ### Equivalence
+We establish the equivalence between `OnePoint K` and `ℙ K (Fin 2 → K)` for a field `K`.
+-/
+
+/-- In a nonzero pair, if one coordinate is 0 then the other is nonzero. -/
+lemma not_both_zero {K : Type} [Zero K]
+    (a : {v : Fin 2 → K // v ≠ 0}) (h : a.1 1 = 0) : a.1 0 ≠ 0 := by
+  intro hc; apply a.2; ext s
+  cases (Nat.le_one_iff_eq_zero_or_eq_one.mp (Fin.is_le s)) with
+  |inl hl =>
+    have : s = 0 := by apply Fin.eq_of_val_eq;tauto
+    subst this;simp_all
+  |inr hr =>
+    have : s = 1 := by apply Fin.eq_of_val_eq;tauto
+    subst this;simp_all
+
+instance {K : Type} [DivisionRing K] : Setoid ({v : Fin 2 → K // v ≠ 0}) :=
+  projectivizationSetoid K (Fin 2 → K)
+
+/-- `divSlope` respects projective equivalence. -/
+lemma divSlope_inj_lifted {K : Type} [Field K]
+    (a b : {v : Fin 2 → K // v ≠ 0}) :
+    divSlope ⟦a⟧ = divSlope ⟦b⟧ →
+    (⟦a⟧ : Quotient (projectivizationSetoid K (Fin 2 → K))) = ⟦b⟧ := by
+  unfold divSlope
+  intro h
+  repeat rw [Quotient.lift_mk] at h
+  apply Quotient.sound
+  unfold divOnePoint at h
+  split_ifs at h with g₀ g₁ g₂
+  · use Units.mk ((a.1 1) / (b.1 1)) ((b.1 1) / (a.1 1)) (by field_simp) (by field_simp)
+    ext s
+    cases (Nat.le_one_iff_eq_zero_or_eq_one.mp (Fin.is_le s)) with
+    |inl hl =>
+      have : s = 0 := Fin.eq_of_val_eq hl
+      subst this; simp_all only [ne_eq, Fin.val_zero, Pi.smul_apply, Units.smul_mk_apply,
+        smul_eq_mul];field_simp
+      have h' : (a.1 0 / a.1 1) = (b.1 0 / b.1 1) := by
+        apply Option.some_injective; tauto
+      field_simp at h';rw [h',mul_comm]
```

Comment:
A style nitpick: mathlib style puts a space after comma and semicolon, so this would be `field_simp at h'; rw [h', mul_comm]`.
(Also, semicolons are used sparingly: tactic calls *can* be put on the same line if they really logically belong together. But this should be intentional.)